AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/DefaultKeyStorageInterface.dfy

```diff
@@ -708,6 +773,522 @@ module DefaultKeyStorageInterface {
+      );
+      var ddbResponse? := ddbClient.Query(ddbRequest);
+
+      /** Handle DDB Error */
+      var ddbResponse :- ddbResponse?
+      .MapFailure(
+        (e: DDB.Error) => wrapDdbException(
+            e:=e,
+            storageOperation:="QueryForVersions",
+            ddbOperation:="Query",
+            identifier:=input.Identifier,
+            tableName:=ddbTableName));
+
+      /** Process sensical DDB Response */
+      var lastKeyBlob: seq<Types.UInt.uint8> := [];
+      var lastKeyEmpty: bool :=
+        // It is not clear if SDKs/Smithy-Dafny/Custom Implementations of Storage will respond with None or an Empty Map.
+        ddbResponse.LastEvaluatedKey.None? || (ddbResponse.LastEvaluatedKey.Some? && |ddbResponse.LastEvaluatedKey.value| == 0);
+
+      if (!lastKeyEmpty) {
+        lastKeyBlob :- lastEvaluatedKeyToBlob(ddbResponse.LastEvaluatedKey.value);
+      }
+      if (ddbResponse.Items.None? || ( |ddbResponse.Items.value| == 0) ) {
+        return Success(Types.QueryForVersionsOutput(
+                         exclusiveStartKey := lastKeyBlob,
+                         items := []
+                       ));
+      }
+
+      var items: seq<Types.EncryptedHierarchicalKey> :- Seq.MapWithResult(
+        (item: DDB.AttributeMap)
+        =>
+          var key :- encryptedHierarchicalKeyFromItem(item, logicalKeyStoreName, input.Identifier);
+          :- Need(
+               key.Type.HierarchicalSymmetricVersion?,
+               Types.KeyStorageException(
+                 message:="Unexpected item returned by DDB. TableName: " + ddbTableName + "\tBranch Key ID: " + input.Identifier
+               ));
+          Success(key),
+        ddbResponse.Items.value
+      );
+
+      return Success(
+          Types.QueryForVersionsOutput(
+            exclusiveStartKey := lastKeyBlob,
+            items := items
+          ));
+    }
+
+    function method blobToExclusiveStartKey(
+      blob: seq<Types.UInt.uint8>,
+      identifier: string
+    ): (output: Result<DDB.Key,Types.Error>)
+    {
+      // From DDB's Docs:
+      // https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#API_Query_RequestSyntax
+      // > The primary key of the first item that this
+      // > operation will evaluate. Use the value that was returned for LastEvaluatedKey in the previous operation.
+      // > The data type for ExclusiveStartKey must be String, Number, or Binary. No set data types are allowed.
+      // From that, we can infer that Partition Key is just going to be the Identifier.
+      // Thus, we only need to store the Type value.
+      // This will be the full "branch:version:<uuidv4>"
+      var versionStr :- UTF8.Decode(blob)
+                        .MapFailure(
+                          eString
+                          =>
+                            Types.KeyStorageException(
+                              message:="Could not UTF8 Decode Exclusive Start Key. " + eString));
```

Comment:
Sorry... I do not get a Dafny Syntax error;
this appears to be correct Dafny code.

Are you saying this not correctly format?
i.e: It is stylistically weird?

In which case,
I wholly agree with you.

But the Dafny formatter wants it this way.

It, in my opinion, in correctly over indents if logical symbols are on the same line.
Hence, `.MapFailure` starts 2 spaces over from where the logical end of UTF8.Decode.

If I had put `versionStr`, `:-`, and `UTF8.Decode` each on their own line,
each would get two additional spaces intended in.

But because I put them all on the same line,
`.MapFailure` starts two spaces from the start of `UTF8.Decode`.

Which... is... the Dafny formatter just wants a lot of vertical space,
and if you do not give it to it,
it will cost you 10 fold  in horizontal space.

I am not a fan.

Reply:
OK. I have another way of formatting this.